nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/PasswordBasedOauth2TokenProvider.java

```diff
@@ -0,0 +1,230 @@
+            SSLContext sslContext = sslService.createContext();
+            clientBuilder.sslSocketFactory(sslContext.getSocketFactory(), trustManager);
+        }
+
+        return clientBuilder.build();
+    }
+
+    @Override
+    public String getAccessToken() {
+        String accessToken;
+
+        if (this.accessDetails == null) {
+            acquireAuthorizationDetails();
+        } else if (this.accessDetails.isExpired()) {
+            try {
+                refreshAuthorizationDetails();
+            } catch (Exception e) {
+                getLogger().info("Couldn't refresh access token, probably refresh token has expired." +
+                    " Getting new token using credentials.");
+                acquireAuthorizationDetails();
+            }
+        }
+        accessToken = accessDetails.getAccessToken();
+
+        return accessToken;
+    }
+
+    private void acquireAuthorizationDetails() {
+        FormBody.Builder acquireTokenBuilder = new FormBody.Builder()
+            .add("grant_type", "password")
+            .add("username", username)
+            .add("password", password);
+
+        if (clientId != null) {
+            acquireTokenBuilder.add("client_id", clientId);
+            acquireTokenBuilder.add("client_secret", clientSecret);
+        }
+
+        RequestBody acquireTokenRequestBody = acquireTokenBuilder.build();
+
+        Request acquireTokenRequest = new Request.Builder()
+            .url(authorizationServerUrl)
+            .post(acquireTokenRequestBody)
+            .build();
+
+        this.accessDetails = getAccessDetails(acquireTokenRequest);
+    }
+
+    private void refreshAuthorizationDetails() {
+        FormBody.Builder refreshTokenBuilder = new FormBody.Builder()
+            .add("grant_type", "refresh_token")
+            .add("refresh_token", this.accessDetails.getRefreshToken());
+
+        if (clientId != null) {
+            refreshTokenBuilder.add("client_id", clientId);
+            refreshTokenBuilder.add("client_secret", clientSecret);
+        }
+
+        RequestBody refreshTokenRequestBody = refreshTokenBuilder.build();
+
+        Request refreshRequest = new Request.Builder()
+                .url(authorizationServerUrl)
+                .post(refreshTokenRequestBody)
+                .build();
+
+        this.accessDetails = getAccessDetails(refreshRequest);
+    }
+
+    private AccessDetails getAccessDetails(Request newRequest) {
+        try {
+            Response response = httpClient.newCall(newRequest).execute();
+            String responseBody = response.body().string();
+            if (response.code() != 200) {
+                getLogger().error(String.format("Bad response from the server during oauth2 request:\n%s", responseBody));
+                throw new ProcessException(String.format("Got HTTP %d during oauth2 request.", response.code()));
+            }
+
+            ObjectMapper mapper = new ObjectMapper()
```

Comment:
The ObjectMapper instance can be reused, instead of being created in every method invocation.